pinot-common/src/main/java/org/apache/pinot/common/metrics/ServerGauge.java

```diff
@@ -45,7 +45,10 @@ public enum ServerGauge implements AbstractMetrics.Gauge {
   // Dedup metrics
   DEDUP_PRIMARY_KEYS_COUNT("dedupPrimaryKeysCount", false),
   CONSUMPTION_QUOTA_UTILIZATION("ratio", false),
-  JVM_HEAP_USED_BYTES("bytes", true);
+  JVM_HEAP_USED_BYTES("bytes", true),
+  // Consumer lag metrics
+  LLC_RECORDS_LAG("recordCount", false),
+  LLC_AVAILABILITY_LAG_MS("availabilityLagMs", false);
```

Comment:
Does this measure time lag? What is availability lag?

Reply:
Does this take care of the lag when a server is building segments or in HOLD state awaiting segment completion?

Reply:
No. We are not handling that. This is purely to show how far behind our processing from input stream is, irrespective of what the engine is doing.

Reply:
But it does not show that. Each time the consumer does not poll the stream, it is that much farther away from freshness. Unless the stream is idle, but then we are not too worried about that right? The idea is to know how stale the data is when queries are being served in a busy production environment.

Reply:
>show how far behind our consumer is from the input stream irrespective of what the engine is doing 

my bad. I corrected my typo above. It is how far our consumer is with respect to the last processed message. 

> The idea is to know how stale the data is when queries are being served in a busy production environment.

The idea is simply to track the progress the consumer is making regardless of what the engine is doing. So, yes, it will increase when we are not polling the stream. that is expected and that is what we want to know. 


> Does this take care of the lag when a server is building segments or in HOLD state awaiting segment completion?

Can you explain what do you mean by this? availability lag should increase when the server is not polling and is busy with segment building/committing.

Reply:
Does the metric show the increase, is the question. Correct me if I am wrong, but the metric will only show the lag when the partition  last ingested events. When segment build starts, the metric should increase in lag, and I don't think that is being handed in the code that you have.

Reply:
We had experimented with similar approach about a month back internally and decided against having a lag metric for partition and come up instead with [this](https://github.com/apache/pinot/pull/9994) approach. Among the motivation we have were included the explosion of metrics if we do one metric for partition. We also address issues like aging of time lag measures and handling of events that affect partition hosting. These extra items do make sense for time based but not for offset based so I do think we should keep offset and time based lag separate. May be focus this pr on offset and the other one in lag. I could modify my PR to also report the partition creating the largest lag if that is something you think it will be useful. We should also consider whether we want to combine the offset metric on a single one: I think we can reuse the tracker object I created on my PR and disable aging.

Reply:
@navina Navina, I sent an email invite for Monday to see if we can converge on these PRs, please forward as required on your side. Thanks!

Reply:
@jugomezv :

> explosion of metrics if we do one metric for partition. 

can you clarify why adding this partition-level metric would cause metric explosion when the realtime engine already has a bunch of per-partition-level metric (eg. `HIGHEST_STREAM_OFFSET_CONSUMED` , `LLC_PARTITION_CONSUMING`, `UPSERT_PRIMARY_KEYS_COUNT` etc )? 

> These extra items do make sense for time based but not for offset based so I do think we should keep offset and time based lag separate. May be focus this pr on offset and the other one in lag. 

This PR does not do offset-based lag as it is pretty expensive. The initial version of the PR did include the offset lag. but there was a flag that made sure it doesn't actually emit the metric. I have updated the PR since to only publish time-based lag. 

> We also address issues like aging of time lag measures and handling of events that affect partition hosting.

By "aging of time lag measure", do you mean incrementing the lag when we are not polling for messages? That would only make the lag trend-up when: 1. there is no data in the input stream  or 2. consumption rate is lower than the ingestion rate . From a monitoring stand-point, it is hard to define a rule that would not result in false positives.

Reply:
Our argument to not have a per partition metric is that it just clouds the user view, what we care about in most cases is the worst case scenario which is what we are handling. Our SREs complained about the per metrica approach as it would make their job harder, they wanted a single point metric that tells them the worst case scenario.

Ack on this having removed the support for offset, but I think that was important.

As for aging, imagine you collect one sample that says event delay last measure was 100ms and then we do get stuck committing a segment, that may take several seconds so the best estimate of consumption delay is the age of the last worst sample plus the sample itself. If we keep posting the sample we are giving the wrong idea to the customer. When we do try to consume and no events are found we display 0 delay as if there were events we will consume them with no delay. Please let me know if the meeting I set for next week suits your schedule and if not let me know what time will be best.

Reply:
> Our SREs complained about the per metrics approach as it would make their job harder, they wanted a single point metric that tells them the worst case scenario.

this is a good point. But we and some users have often found the need to look at per-partition level metrics during debugging. This is esp. true for non-kafka sources for which pinot provides no equivalent lag metrics.

> Ack on this having removed the support for offset, but I think that was important.

In most practical scenarios, offset difference doesn't tell you much. But it can be useful during debugging and this is definitely one of the widely asked/popular metric. However, computing this in Pinot layer is quite expensive as it requires sending a request to the source kafka broker each time. Instead, the kafka consumer client provides APIs to fetch this metric that it holds internally. I think going forward, we should leverage that API or directly try to export the kafka consumer client's mbean which contains the lag. 
Moreover, the offset lag doesn't apply to non-kafka sources anyway. 

> When we do try to consume and no events are found we display 0 delay as if there were events we will consume them with no delay

That's not what this PR is doing though. When there are no events, the delay remains the same as the last sample. This along with the consumption rate should tell you whether there were any new events or not. 

> Please let me know if the meeting I set for next week suits your schedule and if not let me know what time will be best.

I don't see an invite in any of my emails. I can DM you my email in the OSS slack.